sim/battle.ts

```diff
@@ -1859,17 +1859,21 @@ export class Battle {
 			if (source && effect.effectType === 'Move') source.lastDamage = targetDamage;
 
 			const name = effect.fullname === 'tox' ? 'psn' : effect.fullname;
+
 			switch (effect.id) {
 			case 'partiallytrapped':
+				if (target.hitSelf) target.hitSelf = false;
 				this.add('-damage', target, target.getHealth, '[from] ' + this.effectState.sourceEffect.fullname, '[partiallytrapped]');
 				break;
 			case 'powder':
+				if (target.hitSelf) target.hitSelf = false;
 				this.add('-damage', target, target.getHealth, '[silent]');
 				break;
 			case 'confused':
 				this.add('-damage', target, target.getHealth, '[from] confusion');
 				break;
 			default:
+				if (target.hitSelf) target.hitSelf = false;
```

Comment:
I'm still unclear on why you can't just pass `effect` to the `runEvent` you have below, then check `effect.id` in each handler as Marty said.

Even if you want to go this route, it'd be a lot DRYer to:
- Declare a variable and initialize it to `false`
- Set it to `true` only in the `confused` case
- Set `target.hitSelf` to this value at the end

Then you don't even need to change `getConfusionDamage`.

Reply:
some berries don't use onTakeDamage, such as enigma and custap. Does `this.effect.id` work the same as `effect.id` as the parameter or is that different